# Literature assistant - Rag evaluation pipeline

## Why ground truth?

To evaluate retrieval quality we need to know — for each question, which chunk contains the correct answer. This is the ground truth.

We use the A→Q approach:
1. Take each chunk from the database (A = answer/context)
2. Ask the LLM to generate questions that this chunk answers (Q = questions)
3. Store the question together with the chunk ID

Later in `04_evaluation-exp.ipynb`, we use this ground truth to evaluate retrieval:
- Run search for each question
- Check if the original chunk appears in the results
- Calculate Hit Rate and MRR

## Why use an LLM to generate questions?

Writing questions manually for 1612 chunks would be extremely time consuming. The LLM generates realistic questions that a researcher might actually ask, making the evaluation more meaningful.

**Cost note:** Generating questions for all 1612 chunks costs approximately $0.50-1.00 with GPT-4o-mini. Ground truth only needs to be generated once and is saved to `data/ground_truth.csv`.

In [5]:
### Step 1. Connect to the database

Before connecting to the database, make sure docker is running by executing docker start pgvector-papers in the command line.

In [6]:
import psycopg
conn = psycopg.connect(
    "postgresql://user:pswd@localhost:5432/papers"
)
print("Connected!")

records = conn.execute("SELECT COUNT(*) FROM chunks").fetchone()
print(f"The database with {records} records is available now.")

Connected!
The database with (1486,) records is available now.


### Step 2. Fetch the chunks

In [7]:
rows = conn.execute("SELECT id, text, filename, author, year FROM chunks").fetchall()
chunks = [
    {"id": r[0], "text": r[1], "filename": r[2], "author": r[3], "year": r[4]}
    for r in rows
]

print(f"Loaded {len(chunks)} chunks")

Loaded 1486 chunks


### Step 3. Generate ground truth

In [8]:
import json
import pandas as pd
from dotenv import load_dotenv
from openai import OpenAI
from tqdm.auto import tqdm

openai_client = OpenAI()

prompt_template = """
You are generating evaluation questions for a RAG system based on scientific papers.

For the following chunk of text, generate 5 questions that this chunk answers.
Return ONLY a JSON list, no markdown, no code blocks, no explanation.

Example output:
["Question 1?", "Question 2?", "Question 3?"]

Chunk:
{text}

Author: {author}
Year: {year}
""".strip()

results = []

for chunk in tqdm(chunks):
    prompt = prompt_template.format(
        text=chunk["text"],
        author=chunk["author"],
        year=chunk["year"]
    )
    
    response = openai_client.responses.create(
        model="gpt-4o-mini",
        input=[{"role": "user", "content": prompt}]
    )
    
    questions = json.loads(response.output_text)
    
    for q in questions:
        results.append({
            "question": q,
            "chunk_id": chunk["id"],
            "filename": chunk["filename"],
            "author": chunk["author"],
            "year": chunk["year"]
        })

df_questions = pd.DataFrame(results)
df_questions.to_csv("data/ground_truth.csv", index=False)
print(f"Generated {len(df_questions)} questions")

100%|██████████| 1486/1486 [50:33<00:00,  2.04s/it] 

Generated 7430 questions


## Result

Generated ~7430 questions (5 per chunk) from 1486 chunks across 12 papers. Saved to `data/ground_truth.csv`.

These questions are used in below as well as in:
- `04_evaluation-exp.ipynb` — retrieval evaluation
- `05_evaluation-rag.ipynb` — RAG answer quality evaluation

### Step 4. Load the ground truth

In [9]:
df_question = pd.read_csv("data/ground_truth.csv")
ground_truth_retrieval = df_question.to_dict(orient="records")

### Step 5. Evaluate retrieval

In [14]:
from ragbase import RAGBase
from sentence_transformers import SentenceTransformer
from evaluation import evaluate

embedding_model = SentenceTransformer("pritamdeka/S-PubMedBert-MS-MARCO")


assistant = RAGBase(
    conn=conn,
    model=embedding_model,
    llm_client=openai_client,
    llm_model="gpt-4o-mini"
)

evaluate(ground_truth_retrieval, assistant.search)

100%|██████████| 7430/7430 [12:52<00:00,  9.62it/s]


{'hit_rate': 0.6041722745625842, 'mrr': 0.37999129440919477}